community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/SchemaRuleSerialization.java

```diff
@@ -0,0 +1,205 @@
+        {
+        case INDEX_RULE:
+            return readIndexRule( id, source );
+        default:
+            throw new UnsupportedOperationException( format( "Got unknown schema rule type '%d'.", schemaRuleType ) );
+        }
+    }
+
+    private static IndexRule readIndexRule( long id, ByteBuffer source ) throws MalformedSchemaRuleException
+    {
+        SchemaIndexProvider.Descriptor indexProvider = readIndexProviderDescriptor( source );
+        byte indexRuleType = source.get();
+        switch ( indexRuleType )
+        {
+        case GENERAL_INDEX:
+            return IndexRule.indexRule(
+                    id,
+                    NewIndexDescriptorFactory.forSchema( readLabelSchema( source ) ),
+                    indexProvider
+                );
+
+        case UNIQUE_INDEX:
+            long owningConstraint = source.getLong();
+            return IndexRule.constraintIndexRule(
+                    id,
+                    NewIndexDescriptorFactory.uniqueForSchema( readLabelSchema( source ) ),
+                    indexProvider,
+                    owningConstraint
+                );
+
+        default:
+            throw new UnsupportedOperationException( format( "Got unknown index rule type '%d'.", indexRuleType ) );
+        }
+    }
+
+    private static LabelSchemaDescriptor readLabelSchema( ByteBuffer source ) throws MalformedSchemaRuleException
+    {
+        SchemaDescriptor schemaDescriptor = readSchema( source );
+        if ( !(schemaDescriptor instanceof LabelSchemaDescriptor) )
+        {
+            throw new MalformedSchemaRuleException( "IndexRules must have LabelSchemaDescriptors, got " +
+                    ""+schemaDescriptor.getClass().getSimpleName() );
+        }
+        return (LabelSchemaDescriptor)schemaDescriptor;
+    }
+
+    private static SchemaDescriptor readSchema( ByteBuffer source )
+    {
+        byte schemaDescriptorType = source.get();
+        switch ( schemaDescriptorType )
+        {
+        case SIMPLE_LABEL:
+            return SchemaDescriptorFactory.forLabel( source.getInt(), readPropertyIds( source ) );
+        default:
+            throw new UnsupportedOperationException( format( "Got unknown schema descriptor type '%d'.",
+                    schemaDescriptorType ) );
+        }
+    }
+
+    private static int[] readPropertyIds( ByteBuffer source )
+    {
+        short numProperties = source.getShort();
+        int[] propertyIds = new int[numProperties];
+        for ( int i = 0; i < numProperties; i++ )
+        {
+            propertyIds[i] = source.getInt();
+        }
+        return propertyIds;
+    }
+
+    private static SchemaIndexProvider.Descriptor readIndexProviderDescriptor( ByteBuffer source )
+    {
+        String providerKey = getDecodedStringFrom( source );
+        String providerVersion = getDecodedStringFrom( source );
+        return new SchemaIndexProvider.Descriptor( providerKey, providerVersion );
+    }
```

Comment:
I have fixed this, but for some reason a rebase has moved the line so the comment doesn't show as outdated. Dunno what happened there.